File: CodeWars/Mathematics/SquareIntoSquares/SquareIntoSquares.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <math.h>

char* aux(long long num, long long rac) {
    if (num == 0)
        return "[]";

    long long i = rac;
    char* result = malloc(200);
    strcpy(result, "[]");

    while (i >= sqrt(num/2.0)+1) {
        long long diff = num - i*i;
        rac = sqrt(diff);

        if (strcmp(result, "[]") != 0) {

        }
    }

    return result;
}

char* decompose(long long n) {
    return aux(n*n, (sqrt(n*n-1)));
}

void testing(long long n, char* sexpr) {
    char* sact = decompose(n);
    if(strcmp(sact, sexpr) != 0)
        printf("Error. Expected %s but got %s\n", sexpr, sact);
    assert(strcmp(sact, sexpr) == 0);
    if (strcmp(sact, "[]") != 0)
        free(sact); sact = NULL;
}

int main() {

    printf("Testing decompose\n");
    {
        testing(50, "[1,3,5,8,49]");
        testing(44, "[2,3,5,7,43]");
        testing(625, "[2,5,8,34,624]");
        testing(2, "[]");
    }

}

File: CodeWars/Mathematics/SquareIntoSquares/SquareIntoSquares.rs

fn aux(num: i64, mut rac: i64) -> Option<Vec<i64>> {
    if num == 0 {
        return Some(vec![]);
    }

    let mut i = rac;
    let result = None;

    while i >= (num as f64 / 2.0).sqrt() as i64 + 1 {
        let diff = num - i.pow(2);
        rac = (diff as f64).sqrt() as i64;
        let temp_result = aux(diff, rac);

        match temp_result {
            Some(mut x) => {
                x.push(i);
                return Some(x);
            }
            None => (),
        }
        i -= 1;
    }

    result
}

fn decompose(n: i64) -> Option<Vec<i64>> {
    aux(n.pow(2), ((n.pow(2)-1) as f64).sqrt() as i64)
}

fn testing(n: i64, exp: Option<Vec<i64>>) -> () {
    assert_eq!(decompose(n), exp)
}

fn main() {
    testing(50, Some(vec![1,3,5,8,49]));
    testing(44, Some(vec![2,3,5,7,43]));
    testing(625, Some(vec![2,5,8,34,624]));
    testing(5, Some(vec![3,4]));
}
